core/src/main/scala/scalaz/Name.scala

```diff
@@ -52,13 +52,31 @@ object Name {
   }
 }
 
-object Need {
+/** Synchronized, thread-safe version of `Need` */
+object ThreadSafeNeed {
   def apply[A](a: => A): Need[A] = {
     new Need[A] {
-      private[this] lazy val value0: A = a
-      def value = value0
+      lazy val value: A = a
```

Comment:
This test passes, without the `private [this]`:

https://github.com/scalaz/scalaz/pull/1145/files#diff-2edaebab281ed9a80e0bfd3ea3d54db8R5